src/java.base/share/classes/com/sun/crypto/provider/SHA3Parallel.java

```diff
@@ -0,0 +1,127 @@
+
+public class SHA3Parallel {
+    private int blockSize = 0;
+    private static final int WIDTH = 200; // in bytes, e.g. 1600 bits
+    private static final int DM = 5; // dimension of lanesArr
+    private static final int NR = 24; // number of rounds
+
+    // precomputed round constants needed by the step mapping Iota
+    private static final long[] RC_CONSTANTS = {
+            0x01L, 0x8082L, 0x800000000000808aL,
+            0x8000000080008000L, 0x808bL, 0x80000001L,
+            0x8000000080008081L, 0x8000000000008009L, 0x8aL,
+            0x88L, 0x80008009L, 0x8000000aL,
+            0x8000808bL, 0x800000000000008bL, 0x8000000000008089L,
+            0x8000000000008003L, 0x8000000000008002L, 0x8000000000000080L,
+            0x800aL, 0x800000008000000aL, 0x8000000080008081L,
+            0x8000000000008080L, 0x80000001L, 0x8000000080008008L,
+    };
+    private byte[][] buffers;
+    private long[][] lanesArr;
+    private long[] fakeLanes = new long[DM * DM];
+    private int nrPar;
+
+    private SHA3Parallel(byte[][] buffers, int blockSize) {
+        nrPar = buffers.length;
+        this.buffers = buffers;
+        this.blockSize = blockSize;
+        lanesArr = new long[nrPar][];
+        for (int i = 0; i < nrPar; i++) {
+            lanesArr[i] = new long[DM * DM];
+            b2lLittle(buffers[i], 0, lanesArr[i], 0, blockSize);
+        }
+    }
+
+    public void reset(byte[][] buffers) {
+        nrPar = buffers.length;
+        this.buffers = buffers;
+        boolean newSize = (nrPar > lanesArr.length);
+        if (newSize) {
+            lanesArr = new long[nrPar][];
+        }
+        for (int i = 0; i < nrPar; i++) {
+            if (newSize) {
+                lanesArr[i] = new long[DM * DM];
+            } else {
+                Arrays.fill(lanesArr[i], 0L);
+            }
+            b2lLittle(buffers[i], 0, lanesArr[i], 0, blockSize);
+        }
+    }
+
+    public int squeezeBlock() {
+        int retVal = parKeccak();
+        for (int i = 0; i < nrPar; i++) {
+            l2bLittle(lanesArr[i], 0, buffers[i], 0, blockSize);
+        }
+        return retVal;
+    }
+
+    private int parKeccak() {
+        int inlined = 0;
+        for (int i = 0; i < (nrPar + 1) / 2; i ++) {
+            inlined = doubleKeccak(lanesArr[2 * i],
+                    2 * i + 1 == nrPar ? fakeLanes : lanesArr[2 * i + 1]);
+        }
+        return inlined;
+    }
+
+    @IntrinsicCandidate
+    private static int doubleKeccak(long[] lanes0, long[] lanes1) {
+        doubleKeccakJava(lanes0, lanes1);
+        return 1;
+    }
+
+    private static int doubleKeccakJava(long[] lanes0, long[] lanes1) {
+        keccak(lanes0);
+        keccak(lanes1);
+        return 1;
+    }
+
+    public static final class Shake128Parallel extends SHA3Parallel {
+        public Shake128Parallel(byte[][] buf) {
+            super(buf, 168);
+        }
+    }
+
+    public static final class Shake256Parallel extends SHA3Parallel {
```

Comment:
I didn't find usage of this class in this PR? Is this for future usage?

Reply:
This is used in ML-DSA, but I can remove it from this PR